### ISSUE TITLE
2.0 development

### ISSUE THREAD

**Issue by pitr-ch** [ruby-concurrency/concurrent-ruby#520]
How do we develop 2.0? We've discussed several ideas how to clear the code up. When we start it won't be compatible with 1.x.y. Few options come to mind:

* To branch 1.0 and start reorganising and cleaning things on master.
* Or to create second namespace `Concurrent2` in a different directory and start moving things s which are ready there.
* other option?

**Comment by lucasallan** [comment]
:+1: to  `branch 1.0 ` option.

**Comment by pitr-ch** [comment]
There is another thing to consider. 1.0 will live for long time (because of rails 5) and if we release 2.0 before next major rails version we will put users in bad situation. Even if it's timed with next rails. Users may get easily into situation where they would need 1 and 2 both (un-migrated gems on rails 6, new gems using v2 on rails 5, etc.). To allow both versions to be installed we could consider (unusual) approach: release each major version as a new gem `concurrent-ruby-2` and create new namespace `Concurrent2`.

**Comment by rthbound** [comment]
@pitr-ch : I've handled this in the past (different situation, API client gem) using composition. Concurrent::V1::(all), then Concurrent::V2::(all) < Concurrent::V1::(all).

The concurrency classes in the V2 namespace _don't have to_ inherit from V1 classes. That's just something we do when none or few things need to change about a class from one version to the next.

Usage in application could change slightly:
```
#  this setting should be thread safe so one thread can use V1 while others use V2
Concurrent::Config.version = 2 # or 1
Concurrent::Promise.execute { ... }
```

Or you can be explicit -- `Concurrent::V2::Promise.execute { ... }`

WDY?

**Comment by jdantonio** [comment]
Let's create a branch for 2.0 development, abandon `autoload`, refactor within the 2.0 branch as though we *won't* coexist with 1.x, take our time with 2.0 development, and defer figuring out how to make 1.x and 2.x play nicely.

Leading up to 1.0 we had a small community. 1.0 was downloaded 1.5 million times in its first three months. We now have an opportunity to get feedback from a significantly expanded user base and significantly increased production usage. Since we'll be supporting 1.0 for a long time there is not need to rush to 2.0. We can take our time, gather feedback, experiment, and be deliberate. I don't see any reason to release 2.0 during 2016.

That being said, several thoughts (in no particular order):

* There are a number of Edge features that do not depend on the 2.x API (channels, lock-free collections) that we may release with 1.0 core.
  - Some of these features may also allow us to refactor existing 1.x features (highly performant channels, for example).
* I really don't like the idea of a `-2` gem or a `Concurrent2` namespace. If we keep thinking about this I'm sure we'll develop other ideas.
  - In a perfect world we'll have 2.0 features under the main `Concurrent` namespace and we'll have a "special" mode when 1.x and 2.x need to coexist.
  - The `Concurrent::Config.version` suggestion is interesting. We should keep that in mind as we brainstorm and explore alternatives.
* The more I think about `autoload` the more I dislike it.
  - Originally we allowed users to require specific subsets of the gem (for example, `require 'concurrent/future`). We only began using `autoload` when we merged in other gems that already used it.
  - We moved away from subset requires only because of circular references we introduced with the global config.
  - Since them we've cleaned things up and subset requires are possible again. 
  - In 2.x we're going to simplify the global config and fix the circular requires so the problem will solve itself.
  - Subset requires are explicit whereas `autoload` is implicit. The former fits better with our values.
  - If we build 2.0 so that it fully support subset requires, we can easily add `autoload` later.

**Comment by pitr-ch** [comment]
Bulleted thoughts:
* I do not think `Concurrent::Config.version` will be a good fit for `concurrent-ruby` we are working towards eliminating any global configuration, this would bring one.
* If 2.0 will be in 2017 then I would definitely like to get new futures in the core. It could be made compatible and merged in for 1.1 deprecating the classes it replaces.
* We should clear up Concurrent space to avoid collisions as much as possible. E.g. methods like `Concurrent.executor`, `Concurrent.future` should be opt-in as well as a `FactoryMethods` module.
* autoload discussion continues in #395

**Comment by jdantonio** [comment]
We will resume this discussion in other issues while I will be creating over the next several days.
